File: src/ashley/detect.py
"""Ashley Project Detector.

Auto-detects the project's tech stack by scanning for configuration
files and patterns. Produces a context dict that Jinja2 templates
can use for conditional content inclusion.

Example template usage in components:

    {% if project.python %}
    Use `ruff format` for formatting.
    {% endif %}

    {% if project.react %}
    Follow React component conventions...
    {% endif %}
"""

import json
from pathlib import Path


def detect_project(cwd: str | Path | None = None) -> dict:
    """Scan the working directory and return a project context dict.

    Returns a dict with boolean flags for detected technologies
    and string values for detected metadata. All keys are safe
    for use in Jinja2 templates.
    """
    if cwd is None:
        cwd = Path.cwd()
    else:
        cwd = Path(cwd)

    ctx: dict = {
        # Language flags
        "python": False,
        "typescript": False,
        "javascript": False,
        "go": False,
        "rust": False,
        "java": False,
        "dart": False,
        "c": False,
        "cpp": False,
        "shell": False,
        # Framework flags
        "react": False,
        "nextjs": False,
        "vue": False,
        "svelte": False,
        "fastapi": False,
        "django": False,
        "flask": False,
        "express": False,
        "flutter": False,
        # Tooling flags
        "docker": False,
        "makefile": False,
        "git": False,
        "ci_github": False,
        "ci_gitlab": False,
        # Package manager
        "npm": False,
        "pnpm": False,
        "yarn": False,
        "uv": False,
        "pip": False,
        "cargo": False,
        # Metadata
        "name": cwd.name,
        "has_tests": False,
        "has_readme": False,
    }

    # Python detection
    if (cwd / "pyproject.toml").is_file():
        ctx["python"] = True
        content = (cwd / "pyproject.toml").read_text()
        if "fastapi" in content.lower():
            ctx["fastapi"] = True
        if "django" in content.lower():
            ctx["django"] = True
        if "flask" in content.lower():
            ctx["flask"] = True
    if (cwd / "setup.py").is_file() or (cwd / "setup.cfg").is_file():
        ctx["python"] = True
    if (cwd / "requirements.txt").is_file():
        ctx["python"] = True
        ctx["pip"] = True

    # uv detection
    if (cwd / "uv.lock").is_file():
        ctx["uv"] = True
        ctx["python"] = True

    # Node.js / TypeScript detection
    pkg_json = cwd / "package.json"
    if pkg_json.is_file():
        ctx["javascript"] = True
        try:
            pkg = json.loads(pkg_json.read_text())
            deps = {
                **pkg.get("dependencies", {}),
                **pkg.get("devDependencies", {}),
            }
            if "typescript" in deps:
                ctx["typescript"] = True
            if "react" in deps or "react-dom" in deps:
                ctx["react"] = True
            if "next" in deps:
                ctx["nextjs"] = True
                ctx["react"] = True
            if "vue" in deps:
                ctx["vue"] = True
            if "svelte" in deps:
                ctx["svelte"] = True
            if "express" in deps:
                ctx["express"] = True
            ctx["name"] = pkg.get("name", ctx["name"])
        except (json.JSONDecodeError, OSError):
            pass

    if (cwd / "tsconfig.json").is_file():
        ctx["typescript"] = True

    # Package manager detection
    if (cwd / "pnpm-lock.yaml").is_file():
        ctx["pnpm"] = True
    elif (cwd / "yarn.lock").is_file():
        ctx["yarn"] = True
    elif (cwd / "package-lock.json").is_file():
        ctx["npm"] = True

    # Go
    if (cwd / "go.mod").is_file():
        ctx["go"] = True

    # Rust
    if (cwd / "Cargo.toml").is_file():
        ctx["rust"] = True
        ctx["cargo"] = True

    # Java
    if (cwd / "pom.xml").is_file() or (cwd / "build.gradle").is_file():
        ctx["java"] = True

    # Dart / Flutter
    if (cwd / "pubspec.yaml").is_file():
        ctx["dart"] = True
        try:
            content = (cwd / "pubspec.yaml").read_text()
            if "flutter" in content:
                ctx["flutter"] = True
        except OSError:
            pass

    # C / C++
    if (cwd / "CMakeLists.txt").is_file():
        ctx["cpp"] = True
    if (cwd / "Makefile").is_file():
        # Could be C/C++ but Makefile is generic
        ctx["makefile"] = True

    # Docker
    if (cwd / "Dockerfile").is_file() or (cwd / "docker-compose.yml").is_file():
        ctx["docker"] = True

    # Git
    if (cwd / ".git").exists():
        ctx["git"] = True

    # CI
    if (cwd / ".github" / "workflows").is_dir():
        ctx["ci_github"] = True
    if (cwd / ".gitlab-ci.yml").is_file():
        ctx["ci_gitlab"] = True

    # Shell scripts
    sh_files = list(cwd.glob("*.sh")) + list(cwd.glob("scripts/*.sh"))
    if sh_files:
        ctx["shell"] = True

    # Tests
    for test_dir in ("tests", "test", "__tests__", "spec"):
        if (cwd / test_dir).is_dir():
            ctx["has_tests"] = True
            break

    # README
    for readme in ("README.md", "README.rst", "README.txt", "README"):
        if (cwd / readme).is_file():
            ctx["has_readme"] = True
            break

    return ctx

File: internal/project/detect.go
// Package project detects the project context consumed by Ashley templates.
package project

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"
)

// Detect returns the same technology flags and metadata as the Python detector.
func Detect(root string) map[string]any {
	result := make(map[string]any)
	for _, key := range strings.Fields("python typescript javascript go rust java dart c cpp shell react nextjs vue svelte fastapi django flask express flutter docker makefile git ci_github ci_gitlab npm pnpm yarn uv pip cargo has_tests has_readme") {
		result[key] = false
	}
	result["name"] = filepath.Base(filepath.Clean(root))
	file := func(name string) bool {
		info, err := os.Stat(filepath.Join(root, name))
		return err == nil && !info.IsDir()
	}
	dir := func(name string) bool {
		info, err := os.Stat(filepath.Join(root, name))
		return err == nil && info.IsDir()
	}
	if file("pyproject.toml") {
		result["python"] = true
		content, _ := os.ReadFile(filepath.Join(root, "pyproject.toml"))
		for _, dep := range []string{"fastapi", "django", "flask"} {
			if strings.Contains(strings.ToLower(string(content)), dep) {
				result[dep] = true
			}
		}
	}
	if file("setup.py") || file("setup.cfg") {
		result["python"] = true
	}
	if file("requirements.txt") {
		result["python"], result["pip"] = true, true
	}
	if file("uv.lock") {
		result["python"], result["uv"] = true, true
	}
	if file("package.json") {
		result["javascript"] = true
		data, _ := os.ReadFile(filepath.Join(root, "package.json"))
		var pkg struct {
			Name            *string        `json:"name"`
			Dependencies    map[string]any `json:"dependencies"`
			DevDependencies map[string]any `json:"devDependencies"`
		}
		if json.Unmarshal(data, &pkg) == nil {
			for _, deps := range []map[string]any{pkg.Dependencies, pkg.DevDependencies} {
				for dep := range deps {
					switch dep {
					case "typescript", "react", "vue", "svelte", "express":
						result[dep] = true
					case "react-dom":
						result["react"] = true
					case "next":
						result["nextjs"], result["react"] = true, true
					}
				}
			}
			if pkg.Name != nil {
				result["name"] = *pkg.Name
			}
		}
	}
	if file("tsconfig.json") {
		result["typescript"] = true
	}
	for _, pm := range []struct{ name, key string }{{"pnpm-lock.yaml", "pnpm"}, {"yarn.lock", "yarn"}, {"package-lock.json", "npm"}} {
		if file(pm.name) {
			result[pm.key] = true
			break
		}
	}
	for name, key := range map[string]string{"go.mod": "go", "CMakeLists.txt": "cpp", "Makefile": "makefile", ".gitlab-ci.yml": "ci_gitlab"} {
		if file(name) {
			result[key] = true
		}
	}
	if file("Cargo.toml") {
		result["rust"], result["cargo"] = true, true
	}
	if file("pom.xml") || file("build.gradle") {
		result["java"] = true
	}
	if file("pubspec.yaml") {
		result["dart"] = true
		content, _ := os.ReadFile(filepath.Join(root, "pubspec.yaml"))
		if strings.Contains(string(content), "flutter") {
			result["flutter"] = true
		}
	}
	result["docker"] = file("Dockerfile") || file("docker-compose.yml")
	_, err := os.Stat(filepath.Join(root, ".git"))
	result["git"] = err == nil
	result["ci_github"] = dir(".github/workflows")
	for _, pattern := range []string{"*.sh", "scripts/*.sh"} {
		matches, _ := filepath.Glob(filepath.Join(root, pattern))
		if len(matches) > 0 {
			result["shell"] = true
		}
	}
	for _, name := range []string{"tests", "test", "__tests__", "spec"} {
		if dir(name) {
			result["has_tests"] = true
		}
	}
	for _, name := range []string{"README.md", "README.rst", "README.txt", "README"} {
		if file(name) {
			result["has_readme"] = true
		}
	}
	return result
}
